File: C/Euclid_algorithm.c
/*
    *Problem ->
        To find Greatest Common Divisor (GCD) of two numbers.
    * Input ->
        a = 4, b= 6;
    *Output ->
            2
    * Input ->
        a = 100, b= 200;
    *Output ->
            100
*/

/* Solution Idea :
        The basic idea is if we have 2 numbers we can with the min one
        because the gcd is not greater than min one.
 */
    // Euclidean Algorithm
    #include <stdio.h>
    // C function for extended Euclidean Algorithm
      /**
        * Efficient Approach (Euclid's Algorithm)
        Time complexity - O(log(min(a,b)));
        Aux Space - O(1)
    */  
    int gcdExtended(int a, int b, int *x, int *y){
        // Base Case
        if (a == 0){
            *x = 0;
            *y = 1;
            return b;
        }
        int x1, y1; // To store results of recursive call
        int gcd = gcdExtended(b%a, a, &x1, &y1);
        // Update x and y using results of recursive
        // call
        *x = y1 - (b/a) * x1;
        *y = x1;
        return gcd;
    }
    /**
        * Efficient Approach (Euclid's Algorithm) Using recursion
        Time complexity - O(log(min(a,b)));
        Aux Space - O(1)
    */   
    // Function to return gcd of a and b
    int gcd(int a, int b)
    {
        if (a == 0)
            return b;
        return gcd(b%a, a);
    }
    int main(){
         int x, y;
        int c = 35, d = 15;
        int g = gcdExtended(c, d, &x, &y);
        printf("gcd(%d, %d) = %d\n", c, d, g);
        int a = 10, b = 15;
        printf("GCD(%d, %d) = %d\n", a, b, gcd(a, b));
        a = 35, b = 10;
        printf("GCD(%d, %d) = %d\n", a, b, gcd(a, b));
        a = 31, b = 2;
        printf("GCD(%d, %d) = %d\n", a, b, gcd(a, b));
        return 0;
    }


   





 
  

File: Java/Euclid_algorithm.java
/*
    *Problem ->
        To find Greatest Common Divisor (GCD) of two numbers.
    * Input ->
        a = 4, b= 6;
    *Output ->
            2
    * Input ->
        a = 100, b= 200;
    *Output ->
            100
*/

/* Solution Idea :
        The basic idea is if we have 2 numbers we can with the min one
        because the gcd is not greater than min one.
 */
import java.util.*;
public class Euclid_algorithm{
    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter first num : ");
        int a = sc.nextInt();
        System.out.print("Enter second num : ");
        int b = sc.nextInt();

        System.out.println("GCD : ");
        System.out.println(gcdNaive(a,b));
        System.out.println(gcdEuclidBasicVersion(a,b));
        System.out.println(gcdEuclidUsingRecursion(a,b));

        sc.close();
    }
    /**
        * Naive Approach
        Time complexity - O(min(a,b));
        Aux Space - O(1)
    */
    static int gcdNaive(int a , int b){
        int res = Math.min(a,b);
        while(res>0){
            if(a%res == 0 && b%res == 0)
                break;
            res--;
        }
        return res;
    }
    /**
        * Efficient Approach (Euclid's Algorithm)
        Time complexity - O(log(min(a,b)));
        Aux Space - O(1)
    */  
    static int gcdEuclidBasicVersion(int a, int b){
        while(a != b){
            if(a>b) a = a-b;
            else b = b-a;
        }
        return a;
    } 
    /**
        * Efficient Approach (Euclid's Algorithm) Using recursion
        Time complexity - O(log(min(a,b)));
        Aux Space - O(1)
    */   
    static int gcdEuclidUsingRecursion(int a, int b){
        if(b==0) return a;
        else return gcdEuclidUsingRecursion(b, a%b);
    }



}

 
  